File: backend/server/translate_curated.py
import os
import json
from dotenv import load_dotenv
from supabase import create_client, Client
from google import genai
from pydantic import BaseModel, Field

# Load env
load_dotenv("../.env")

url: str = os.environ.get("SUPABASE_URL")
key: str = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
supabase: Client = create_client(url, key)

client = genai.Client(api_key=os.environ.get("GEMINI_API_KEY"))

class TranslationOutput(BaseModel):
    name: str
    blurb: str = Field(default="")

def translate_text(loc: dict, target_lang: str) -> dict:
    prompt = f"""
    Translate the following location information to {target_lang}.
    It is a curated location in Algeria.
    If the name is a proper noun (e.g. 'Maqam Echahid'), keep it appropriate for the language, or translate if standard.
    
    Name: {loc.get('name')}
    Blurb: {loc.get('blurb', '')}
    
    Return a JSON object with 'name' and 'blurb'.
    """
    
    generation_config = {
        'max_output_tokens': 2048,
        'temperature': 0.1,
        'response_mime_type': 'application/json',
        'response_schema': TranslationOutput
    }
    
    response = client.models.generate_content(
        model='gemini-2.5-flash',
        contents=prompt,
        config=generation_config
    )
    
    return json.loads(response.text)

def main():
    print("Fetching curated locations (English translations)...")
    res = supabase.table("location_translations").select("location_id, name, blurb").eq("locale", "en").execute()
    locations = res.data
    
    for loc in locations:
        print(f"Translating: {loc['name']}")
        for lang in ['fr', 'ar']:
            existing = supabase.table("location_translations").select("location_id").eq("location_id", loc["location_id"]).eq("locale", lang).execute()
            if existing.data:
                print(f"  [{lang}] already exists, skipping.")
                continue
            
            print(f"  [{lang}] Translating...")
            try:
                translated = translate_text(loc, "French" if lang == 'fr' else "Arabic (Modern Standard)")
                supabase.table("location_translations").insert({
                    "location_id": loc["location_id"],
                    "locale": lang,
                    "name": translated.get("name") or loc.get("name"),
                    "blurb": translated.get("blurb")
                }).execute()
                print(f"  [{lang}] Saved.")
            except Exception as e:
                print(f"  [{lang}] Error: {e}")

if __name__ == "__main__":
    main()

File: backend/server-node/src/scripts/translateCurated.ts
/**
 * One-off: translate the curated locations' English blurbs into fr and ar.
 *
 * Skips any (location, locale) pair that already exists, so it is safe to
 * re-run. The Python version declared its output shape with a pydantic model
 * passed as `response_schema`; the JS SDK takes the same thing as a plain
 * JSON-schema object, and the result is parsed the same way either way.
 *
 *     npm run translate:curated
 */
import { GoogleGenAI } from "@google/genai";

import { Config, requireGeminiKey } from "../config";
import { getAdminClient } from "../ingestion/supabaseAdmin";
import { unwrapRows } from "../supabase";

const TRANSLATION_SCHEMA = {
  type: "object",
  properties: {
    name: { type: "string" },
    blurb: { type: "string" },
  },
  required: ["name"],
};

interface TranslationOutput {
  name?: string;
  blurb?: string;
}

interface TranslationRow {
  location_id: string;
  name: string;
  blurb: string;
}

async function translateText(
  ai: GoogleGenAI,
  loc: TranslationRow,
  targetLang: string,
): Promise<TranslationOutput> {
  const prompt = `
    Translate the following location information to ${targetLang}.
    It is a curated location in Algeria.
    If the name is a proper noun (e.g. 'Maqam Echahid'), keep it appropriate for the language, or translate if standard.

    Name: ${loc.name}
    Blurb: ${loc.blurb ?? ""}

    Return a JSON object with 'name' and 'blurb'.
  `;

  const response = await ai.models.generateContent({
    model: "gemini-2.5-flash",
    contents: prompt,
    config: {
      maxOutputTokens: 2048,
      temperature: 0.1,
      responseMimeType: "application/json",
      responseSchema: TRANSLATION_SCHEMA,
    } as never,
  });

  return JSON.parse(response.text ?? "{}") as TranslationOutput;
}

async function main(): Promise<void> {
  if (!Config.SUPABASE_URL) {
    throw new Error("SUPABASE_URL is not set. Add it to backend/.env.");
  }
  const supabase = getAdminClient();
  const ai = new GoogleGenAI({ apiKey: requireGeminiKey() });

  console.log("Fetching curated locations (English translations)...");
  const locations = await unwrapRows<TranslationRow>(
    supabase.from("location_translations").select("location_id, name, blurb").eq("locale", "en"),
  );

  for (const loc of locations) {
    console.log(`Translating: ${loc.name}`);
    for (const lang of ["fr", "ar"]) {
      const existing = await unwrapRows<{ location_id: string }>(
        supabase
          .from("location_translations")
          .select("location_id")
          .eq("location_id", loc.location_id)
          .eq("locale", lang),
      );
      if (existing.length > 0) {
        console.log(`  [${lang}] already exists, skipping.`);
        continue;
      }

      console.log(`  [${lang}] Translating...`);
      try {
        const translated = await translateText(
          ai,
          loc,
          lang === "fr" ? "French" : "Arabic (Modern Standard)",
        );
        const { error } = await supabase.from("location_translations").insert({
          location_id: loc.location_id,
          locale: lang,
          name: translated.name || loc.name,
          blurb: translated.blurb,
        });
        if (error) throw error;
        console.log(`  [${lang}] Saved.`);
      } catch (error) {
        console.log(`  [${lang}] Error: ${error instanceof Error ? error.message : error}`);
      }
    }
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
